lib/Parse/ParseDecl.cpp

```diff
@@ -9013,13 +9013,14 @@ parseDeclDeinit(ParseDeclOptions Flags, DeclAttributes &Attributes) {
 
   DD->getAttrs() = Attributes;
 
-  // Reject 'destructor' functions outside of structs, enums, and classes.
+  // Reject 'destructor' functions outside of structs, enums, classes, and
+  // @objcImplementation extensions.
   //
   // Later in the type checker, we validate that structs/enums only do this if
-  // they are move only.
-  auto *nom = dyn_cast<NominalTypeDecl>(CurDeclContext);
-  if (!nom ||
-      (!isa<StructDecl>(nom) && !isa<EnumDecl>(nom) && !isa<ClassDecl>(nom))) {
+  // they are move only, and that @objcImplementations are main-body.
+  auto *ED = dyn_cast<ExtensionDecl>(CurDeclContext);
+  if (!(isa<StructDecl>(CurDeclContext) || isa<EnumDecl>(CurDeclContext) ||
+        isa<ClassDecl>(CurDeclContext) || (ED && ED->isObjCImplementation()))) {
```

Comment:
**SwiftParser reviewers**: This is the only change I made to the parser; `SwiftParser.Parser.parseDeinitializerDeclaration(_:_:)` does not diagnose this semantic restriction.